File: sdks/node/src/resources/webhooks.ts
/**
 * `client.webhookEndpoints` and `client.webhookDeliveries` — configuring where events go, and
 * seeing what happened when they went (M22.3).
 *
 * ## The one thing to know about `create`
 *
 * It returns the signing secret, and that is the only time the platform will ever send it.
 * `retrieve` and `list` do not include it. That is deliberate on the platform's side and is
 * reflected here by the return type: `create` and `rotateSecret` resolve to a value that has a
 * `signingSecret`, and nothing else does — so "I'll fetch it again later" does not type-check
 * rather than failing at runtime in whatever code was going to store it.
 *
 * Deliveries are offset-paginated rather than cursor-paginated. That is D139's deliberate
 * exception, not an oversight here: `/v1/webhook_deliveries` and `/v1/test/decisions` were
 * left on the older `PageResponse` envelope, and the SDK reports what the endpoint returns.
 */

import { OPERATIONS } from '../generated/operations.js';
import type {
  WebhookDeliveryResponse,
  WebhookEndpointCreatedResponse,
  WebhookEndpointResponse,
} from '../generated/models.js';
import type { RequestOptions } from '../transport.js';
import type { OffsetPage } from '../pagination.js';
import { Resource } from './base.js';

/** What `webhookEndpoints.create` accepts. */
export interface WebhookEndpointCreateParams {
  /** Where to deliver events. Must be HTTPS. */
  readonly url: string;
  /** Which event types this endpoint should receive. */
  readonly enabledEvents: readonly string[];
  /** A description for your own records. */
  readonly description?: string;
  readonly metadata?: Readonly<Record<string, string>>;
}

/** What `webhookEndpoints.update` accepts. Every field optional: send only what changes. */
export interface WebhookEndpointUpdateParams {
  readonly enabled?: boolean;
  readonly enabledEvents?: readonly string[];
  readonly description?: string;
  readonly metadata?: Readonly<Record<string, string>>;
}

export class WebhookEndpoints extends Resource {
  /**
   * Creates an endpoint and returns it **with its signing secret**.
   *
   * Store the secret now. It is not retrievable afterwards — `rotateSecret` issues a new one
   * rather than showing you this one again.
   */
  create(params: WebhookEndpointCreateParams, options?: RequestOptions): Promise<WebhookEndpointCreatedResponse> {
    return this.send<WebhookEndpointCreatedResponse>(OPERATIONS.createWebhookEndpoint, { body: params, options });
  }

  /** Retrieves one endpoint. Never includes the signing secret. */
  retrieve(id: string, options?: RequestOptions): Promise<WebhookEndpointResponse> {
    return this.send<WebhookEndpointResponse>(OPERATIONS.getWebhookEndpoint, { path: { id }, options });
  }

  /**
   * Lists your endpoints.
   *
   * Returns a plain array, because the endpoint does: this list is not paginated on the wire,
   * and wrapping it in a page object would invent a `hasMore` that no response carries.
   */
  list(options?: RequestOptions): Promise<WebhookEndpointResponse[]> {
    return this.send<WebhookEndpointResponse[]>(OPERATIONS.listWebhookEndpoints, { options });
  }

  /** Updates an endpoint. */
  update(
    id: string,
    params: WebhookEndpointUpdateParams,
    options?: RequestOptions,
  ): Promise<WebhookEndpointResponse> {
    return this.send<WebhookEndpointResponse>(OPERATIONS.updateWebhookEndpoint, {
      path: { id },
      body: params,
      options,
    });
  }

  /** Deletes an endpoint. Resolves with nothing: the API returns 204. */
  del(id: string, options?: RequestOptions): Promise<void> {
    return this.send<void>(OPERATIONS.deleteWebhookEndpoint, { path: { id }, options });
  }

  /**
   * Issues a new signing secret for an endpoint and returns it.
   *
   * As with `create`, this is the only time the new secret is sent.
   */
  rotateSecret(id: string, options?: RequestOptions): Promise<WebhookEndpointCreatedResponse> {
    return this.send<WebhookEndpointCreatedResponse>(OPERATIONS.rotateWebhookEndpointSecret, {
      path: { id },
      options,
    });
  }
}

/** What `webhookDeliveries.list` accepts. Offset pagination, per D139. */
export type WebhookDeliveryListParams = {
  /** Zero-based page index. */
  readonly page?: number;
  /** How many deliveries per page. */
  readonly size?: number;
  /** Sort instructions, such as `createdAt,desc`. */
  readonly sort?: readonly string[];
};

export class WebhookDeliveries extends Resource {
  /** Retrieves one delivery, including its attempts. */
  retrieve(id: string, options?: RequestOptions): Promise<WebhookDeliveryResponse> {
    return this.send<WebhookDeliveryResponse>(OPERATIONS.getWebhookDelivery, { path: { id }, options });
  }

  /** Lists deliveries. The result paginates transparently. */
  list(
    params: WebhookDeliveryListParams = {},
    options?: RequestOptions,
  ): Promise<OffsetPage<WebhookDeliveryResponse>> {
    return this.listOffset<WebhookDeliveryResponse>(OPERATIONS.listWebhookDeliveries, params, options);
  }

  /** Re-sends a delivery. Returns the new delivery, not the original. */
  replay(id: string, options?: RequestOptions): Promise<WebhookDeliveryResponse> {
    return this.send<WebhookDeliveryResponse>(OPERATIONS.replayWebhookDelivery, { path: { id }, options });
  }
}

File: sdks/python/src/paymentflow/resources/webhooks.py
"""``client.webhook_endpoints`` and ``client.webhook_deliveries`` — configuring where events go,
and seeing what happened when they went (M22.6).

The one thing to know about ``create``
--------------------------------------
It returns the signing secret, and that is the only time the platform will ever send it.
``retrieve`` and ``list`` do not include it. ``rotate_secret`` issues a *new* one rather than
showing you this one again.

Deliveries are offset-paginated rather than cursor-paginated. That is D139's deliberate
exception, not an oversight here: ``/v1/webhook_deliveries`` and ``/v1/test/decisions`` were
left on the older ``PageResponse`` envelope, and the SDK reports what the endpoint returns.
"""

from __future__ import annotations

from typing import Mapping, Optional, Sequence

from .._generated.models import (
    WebhookDeliveryResponse,
    WebhookEndpointCreatedResponse,
    WebhookEndpointResponse,
)
from .._generated.operations import OPERATIONS
from .._pagination import OffsetPage
from .._transport import RequestOptions
from ._base import Resource, body_of, query_of

__all__ = ["WebhookDeliveries", "WebhookEndpoints"]


class WebhookEndpoints(Resource):
    def create(
        self,
        *,
        url: str,
        enabled_events: Sequence[str],
        description: Optional[str] = None,
        metadata: Optional[Mapping[str, str]] = None,
        options: Optional[RequestOptions] = None,
    ) -> WebhookEndpointCreatedResponse:
        """Creates an endpoint and returns it **with its signing secret**.

        Store the secret now. It is not retrievable afterwards.
        """
        body = body_of(
            url=url,
            enabledEvents=list(enabled_events),
            description=description,
            metadata=metadata,
        )
        result: WebhookEndpointCreatedResponse = self._send(
            OPERATIONS["createWebhookEndpoint"], body=body, options=options
        )
        return result

    def retrieve(self, endpoint_id: str, *, options: Optional[RequestOptions] = None) -> WebhookEndpointResponse:
        """Retrieves one endpoint. Never includes the signing secret."""
        result: WebhookEndpointResponse = self._send(
            OPERATIONS["getWebhookEndpoint"], path={"id": endpoint_id}, options=options
        )
        return result

    def list(self, *, options: Optional[RequestOptions] = None) -> Sequence[WebhookEndpointResponse]:
        """Lists your endpoints.

        Returns a plain list, because the endpoint does: this list is not paginated on the
        wire, and wrapping it in a page object would invent a ``has_more`` that no response
        carries. The set is capped at 16 per mode.
        """
        result: Sequence[WebhookEndpointResponse] = self._send(
            OPERATIONS["listWebhookEndpoints"], options=options
        )
        return result

    def update(
        self,
        endpoint_id: str,
        *,
        enabled: Optional[bool] = None,
        enabled_events: Optional[Sequence[str]] = None,
        description: Optional[str] = None,
        metadata: Optional[Mapping[str, str]] = None,
        options: Optional[RequestOptions] = None,
    ) -> WebhookEndpointResponse:
        """Updates an endpoint. Send only what changes."""
        body = body_of(
            enabled=enabled,
            enabledEvents=None if enabled_events is None else list(enabled_events),
            description=description,
            metadata=metadata,
        )
        result: WebhookEndpointResponse = self._send(
            OPERATIONS["updateWebhookEndpoint"], path={"id": endpoint_id}, body=body, options=options
        )
        return result

    def delete(self, endpoint_id: str, *, options: Optional[RequestOptions] = None) -> None:
        """Deletes an endpoint. Returns nothing: the API returns 204.

        Spelled ``delete`` rather than the Node SDK's ``del``, which is a Python keyword. This
        is the one place the two SDKs' method names differ, and the difference is forced by the
        languages rather than chosen.
        """
        self._send(OPERATIONS["deleteWebhookEndpoint"], path={"id": endpoint_id}, options=options)

    def rotate_secret(
        self, endpoint_id: str, *, options: Optional[RequestOptions] = None
    ) -> WebhookEndpointCreatedResponse:
        """Issues a new signing secret for an endpoint and returns it.

        As with ``create``, this is the only time the new secret is sent.
        """
        result: WebhookEndpointCreatedResponse = self._send(
            OPERATIONS["rotateWebhookEndpointSecret"], path={"id": endpoint_id}, options=options
        )
        return result


class WebhookDeliveries(Resource):
    def retrieve(self, delivery_id: str, *, options: Optional[RequestOptions] = None) -> WebhookDeliveryResponse:
        """Retrieves one delivery, including its attempts."""
        result: WebhookDeliveryResponse = self._send(
            OPERATIONS["getWebhookDelivery"], path={"id": delivery_id}, options=options
        )
        return result

    def list(
        self,
        *,
        page: Optional[int] = None,
        size: Optional[int] = None,
        sort: Optional[Sequence[str]] = None,
        options: Optional[RequestOptions] = None,
    ) -> OffsetPage[WebhookDeliveryResponse]:
        """Lists deliveries. The result paginates transparently.

        ``sort`` takes instructions such as ``"createdAt,desc"``.
        """
        query = query_of(page=page, size=size, sort=None if sort is None else list(sort))
        return self._list_offset(OPERATIONS["listWebhookDeliveries"], query, options)

    def replay(self, delivery_id: str, *, options: Optional[RequestOptions] = None) -> WebhookDeliveryResponse:
        """Re-sends a delivery. Returns the new delivery, not the original."""
        result: WebhookDeliveryResponse = self._send(
            OPERATIONS["replayWebhookDelivery"], path={"id": delivery_id}, options=options
        )
        return result
